cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -152,7 +153,7 @@ controller, and serviceaccounts controller.`,
 			// add feature enablement metrics
 			fg := s.ComponentGlobalsRegistry.FeatureGateFor(utilversion.DefaultKubeComponent)
 			fg.(featuregate.MutableFeatureGate).AddMetrics()
-			return Run(context.Background(), c.Complete())
+			return Run(cmd.Context(), c.Complete())
```

Comment:
FIY https://github.com/kubernetes/kubernetes/pull/125196#discussion_r1623727817
This change does not seem to be necessary. 
Or what benefits will there be?

Reply:
When I start ControllerManager outside the container environment, it cannot be shut down normally, and my automatic deployment will fail, and can be Override by cmd.SetContext(xx), it is more elegant than before
such as i want to start api-server and controller manager in one process, i prefer stop controllerManager safely when api-server stop, but now i can't rewrite ctx

Reply:
After relying on a third-party library, why can it solve the problem you mentioned? I'm not quite clear, I'm sorry.

Reply:
Just like k3s, i want to add controllerManager in my project and can follow the community's version, just like
```go
func main() {
    // my own business
    args := []string{xxxx}
    command := app2.NewControllerManagerCommand(ctx)
    command.SetArgs(args)
    command.Execute()
}

```
Now i have no idea to insert context in kube-CM
But follow this pr i can choose context with default (it just like before, so there is no need for improvement), or use my own context just like
```go
func main() {
    // my own business
   ctx, cancel := context.WithCancel(context.Background())
   go func() {
       // if business failed
       cancel()
    }
    args := []string{xxxx}
    command := app2.NewControllerManagerCommand(ctx)
    command.SetArgs(args)
    command.SetContext(ctx)
    command.Execute()
}
```
And also, it can be safe-close without block.

Reply:
>  i want to start api-server and controller manager in one process

there use to be such model with hyperkube that was completely remove in 1.19 https://github.com/kubernetes/kubernetes/pull/88676

> Just like k3s, i want to add controllerManager in my project and can follow the community's version, just like

Kubernetes components are not made to be consumed, and I don't feel we should create this dependency ... 

How is doing k3s then, can you not follow the same approach?